Completing assignment

tassos opened the pull request:
Hey all,

I just completed (I think) the assignment. Here's what I have done:

# App specifications

- Created a Shop and a ProductType model. ProductType model contains the possible values for the types of products that each shop is selling. Shop model contains all the info for the shops (name, description, opening and closing hours, product types)
- I created a 'new shop' form and an index view for all the shops.
- Created two checkboxes for filtering shops that are open now, open in an hour, or both. It is also possible to remove the filters.

# Technical specifications

- The views are coded in HAML. It is not as difficult as I thought
- I used bootstrap, bootstrap-form and bootstrap-timepicker for the views and forms
- I implemented basic rspec tests using Shoulda-Matchers, after @robikovacs suggestion.

Anything else that you think should be done?

florinionce commented on app/controllers/shops_controller.rb:
```diff
@@ -0,0 +1,50 @@
+		end
+
+		# Sanitise the parameters for the filter
+		def filter_params
```
Nice idea :)

florinionce commented on app/controllers/shops_controller.rb:
```diff
@@ -0,0 +1,50 @@
+				opening_time = (Time.now+1.hour).strftime("%H%M%S") #adding one hour
+				closing_time = opening_time
+			end
+			@shops = Shop.select { |shop| shop.opening.strftime("%H%M%S") <= opening_time && shop.closing.strftime("%H%M%S") > closing_time }
```
Hmm, I didn't know you could use `select` directly on an active record model.

What does the generated sql query looks like? :)

robikovacs replied:
Also is a good idea to extract the select statement into a class method on the shop model: http://api.rubyonrails.org/classes/ActiveRecord/Scoping/Named/ClassMethods.html

tassos replied:
I think it makes something like this:

`
SELECT * FROM Shop
WHERE opening >= a_time; `

florinionce commented on app/views/shops/index.html.haml:
```diff
@@ -0,0 +1,28 @@
+					%td= shop.description
+					%td= shop.opening.strftime("%H:%M")
+					%td= shop.closing.strftime("%H:%M")
+					%td= shop.product_type.name
```
I think this is causing an n+1 query.
Probably this would solve it:
```ruby
@shops = Shop.includes(:product_type)
```
In the index method

tassos replied:
Not sure I understand what you mean by n+1 query. Maybe we should discuss this live

florinionce replied:
Sure. Will do that :)

robikovacs commented on test/controllers/shops_controller_test.rb:
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
```
You can remove test files because you changed to rspec

tassos replied:
+1 I removed them all the minitest files in 2062336

robikovacs commented on test/models/product_type_test.rb:
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
```
You can remove this

tassos replied:
Same as above

robikovacs commented on test/models/shop_test.rb:
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
```
Remove this

tassos replied:
Same as above

robikovacs commented on spec/models/shop_spec.rb:
```diff
@@ -0,0 +1,13 @@
+	it { should validate_presence_of(:description) }
+	it { should validate_presence_of(:opening) }
+	it { should validate_presence_of(:closing) }
+	it { should validate_presence_of(:product_type) }
```
Cool! I told you it was easy.

tassos replied:
Yes indeed, thanks a lot @robikovacs !

robikovacs commented on app/controllers/shops_controller.rb:
```diff
@@ -0,0 +1,50 @@
+class ShopsController < ApplicationController
+	
```
The indentation is weird. I think you used to many spaces/tabs

tassos replied:
I have setup my editor to use tabs by default, but it seems that Rails uses spaces by default. That's why it is a bit weird indeed. @florinionce also pointed out, so I will change to spaces everywhere (and change the settings of my editor)

tassos commented:
Thanks for the feedback guys!

I will respond to the comments in the evening, as i am attending a conference today and i don't have the laptop with me...

But in general I agree with the comments and they do make a lot of sense.

Kind regards,
Tassos

On 5 July 2017 09:52:47 EEST, Robert Kovacs <notifications@github.com> wrote:
>robikovacs approved this pull request.
>
>Great job @tassos ! 🥇 
>
>> +			params.require(:filter).permit(:open_now, :open_in_one_hour)
>+		end
>+
>+		# Restore just the shops that are open on the relevant timeframe
>+		def filter_shops
>+			if filter_params[:open_now] && filter_params[:open_in_one_hour]
>+				opening_time = Time.now.strftime("%H%M%S")
>+				closing_time = (Time.now+1.hour).strftime("%H%M%S") #adding one
>hour
>+			elsif filter_params[:open_now]
>+				opening_time = Time.now.strftime("%H%M%S")
>+				closing_time = opening_time
>+			elsif filter_params[:open_in_one_hour]
>+				opening_time = (Time.now+1.hour).strftime("%H%M%S") #adding one
>hour
>+				closing_time = opening_time
>+			end
>+			@shops = Shop.select { |shop| shop.opening.strftime("%H%M%S") <=
>opening_time && shop.closing.strftime("%H%M%S") > closing_time }
>
>Also is a good idea to extract the select statement into a class method
>on the shop model:
>http://api.rubyonrails.org/classes/ActiveRecord/Scoping/Named/ClassMethods.html
>
>> @@ -0,0 +1,7 @@
>+require 'test_helper'
>
>You can remove test files because you changed to rspec
>
>> @@ -0,0 +1,7 @@
>+require 'test_helper'
>
>You can remove this
>
>> @@ -0,0 +1,7 @@
>+require 'test_helper'
>
>Remove this
>
>> @@ -0,0 +1,13 @@
>+require 'rails_helper'
>+
>+RSpec.describe Shop, type: :model do
>+	# Association tests
>+	it { should belong_to(:product_type) }
>+
>+	# Validation tests
>+	it { should validate_presence_of(:name) }
>+	it { should validate_presence_of(:description) }
>+	it { should validate_presence_of(:opening) }
>+	it { should validate_presence_of(:closing) }
>+	it { should validate_presence_of(:product_type) }
>
>Cool! I told you it was easy.
>
>> @@ -0,0 +1,50 @@
>+class ShopsController < ApplicationController
>+	
>
>The indentation is weird. I think you used to many spaces/tabs
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/tassos/shopify/pull/1#pullrequestreview-47964060

-- 
web: www.natsakis.com
tel: +40 753 820782

tassos commented:
@florinionce : regarding the plaintext credentials, very good point. I changed the configuration so that it uses environmental variables in 335f8bab. I also converted from tabs to spaces in daff20cee.

Hope that is good now :)
